File: misc/StringCompression.cpp
#include <iostream>
#include <string>
#include <map>

using namespace std;

using ll = long long;

string solve(string s) {
    if(s.size() == 0) {
        return "";
    }

    map<char, ll> tmp;

    char prev = s[0];
    tmp[prev] = 1;

    string result = "";

    for(int i = 1; i < s.size(); i++) {
        const char current = s[i];

        if(current == prev) {
            tmp[current]++; 
        } else {
            result += prev + to_string(tmp[prev]);
            tmp[prev] = 0;
            
            tmp[current] = 1;
            prev = current;
        }
    }

    result += prev + to_string(tmp[prev]);

    return (result.size() < s.size())? result : s;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cout << solve("aabcccccaaa") << endl;
    cout << solve("x")           << endl;

    return 0;
}

File: edu/implementations/misc/StringCompression.ts
function compress(s: string): string {
    if(s.length === 0) {
        return "";
    }

    const tmp = new Map<string, number>();
    let prev = s[0];
    tmp.set(prev, 1);

    let result = "";

    for(let i = 1; i < s.length; i++) {
        const current = s[i];

        if(current === prev) {
            tmp.set(prev, tmp.get(prev)! + 1);
        } else {
            result += `${prev}${tmp.get(prev)}`;
            tmp.delete(prev);
            tmp.set(current, 1);
            prev = current;
        }
    }

    result += `${prev}${tmp.get(prev)}`

    return (result.length < s.length)? result : s;
}

console.log(compress("aabcccccaaa"))
console.log(compress("x"))
